File: converter.py
# converter.py
# Версия на Python с использованием dataclasses, enum, argparse, JSON

import sys
import json
import argparse
import os
from enum import Enum
from dataclasses import dataclass
from typing import Dict, Optional, List, Tuple
import readline  # для истории ввода

# ANSI-цвета
class Colors:
    RESET = '\033[0m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BOLD = '\033[1m'

def colorize(text: str, color: str) -> str:
    return f"{color}{text}{Colors.RESET}"

# Категории и единицы
class Category(Enum):
    LENGTH = "length"
    AREA = "area"
    VOLUME = "volume"
    MASS = "mass"
    PRESSURE = "pressure"
    TEMPERATURE = "temperature"
    ANGLE = "angle"
    SPEED = "speed"

# Базовые единицы (система СИ)
UNITS = {
    Category.LENGTH: {
        "m": 1.0, "cm": 0.01, "mm": 0.001, "km": 1000.0,
        "in": 0.0254, "ft": 0.3048, "yd": 0.9144, "mi": 1609.344
    },
    Category.AREA: {
        "m2": 1.0, "cm2": 0.0001, "ft2": 0.09290304, "yd2": 0.83612736,
        "acre": 4046.8564224, "ha": 10000.0
    },
    Category.VOLUME: {
        "m3": 1.0, "cm3": 1e-6, "l": 0.001, "ft3": 0.028316846592,
        "yd3": 0.764554857984, "gal": 0.003785411784
    },
    Category.MASS: {
        "kg": 1.0, "g": 0.001, "mg": 1e-6, "lb": 0.45359237,
        "oz": 0.028349523125, "t": 1000.0
    },
    Category.PRESSURE: {
        "pa": 1.0, "kpa": 1000.0, "mpa": 1e6, "bar": 100000.0,
        "psi": 6894.757293168, "atm": 101325.0
    },
    Category.TEMPERATURE: {
        # особый случай: не линейная шкала, используем функции
        "c": "celsius", "f": "fahrenheit", "k": "kelvin"
    },
    Category.ANGLE: {
        "deg": 1.0, "rad": 57.29577951308232, "grad": 0.9
    },
    Category.SPEED: {
        "ms": 1.0, "kmh": 0.27777777777778, "mph": 0.44704, "kn": 0.51444444444444
    }
}

# Функции для температуры
def temp_convert(value: float, from_unit: str, to_unit: str) -> float:
    # Сначала переводим в Кельвины
    if from_unit == "c":
        kelvin = value + 273.15
    elif from_unit == "f":
        kelvin = (value + 459.67) * 5/9
    elif from_unit == "k":
        kelvin = value
    else:
        raise ValueError(f"Неизвестная единица температуры: {from_unit}")
    # Из Кельвинов в целевое
    if to_unit == "c":
        return kelvin - 273.15
    elif to_unit == "f":
        return kelvin * 9/5 - 459.67
    elif to_unit == "k":
        return kelvin
    else:
        raise ValueError(f"Неизвестная единица температуры: {to_unit}")

@dataclass
class Conversion:
    value: float
    from_unit: str
    to_unit: str
    result: float
    category: Category

class Converter:
    def __init__(self):
        self.history: List[Conversion] = []
        self.config_file = os.path.expanduser("~/.converter_config.json")
        self.load_config()

    def load_config(self):
        if os.path.exists(self.config_file):
            with open(self.config_file, 'r') as f:
                data = json.load(f)
                self.history = [Conversion(**h) for h in data.get('history', [])]

    def save_config(self):
        with open(self.config_file, 'w') as f:
            json.dump({'history': [h.__dict__ for h in self.history]}, f, indent=2)

    def get_category(self, unit: str) -> Optional[Category]:
        for cat, units in UNITS.items():
            if unit in units:
                return cat
            # для температуры
            if cat == Category.TEMPERATURE and unit in UNITS[Category.TEMPERATURE]:
                return cat
        return None

    def convert(self, value: float, from_unit: str, to_unit: str) -> Tuple[float, Category]:
        from_cat = self.get_category(from_unit)
        to_cat = self.get_category(to_unit)
        if from_cat is None or to_cat is None:
            raise ValueError("Неизвестная единица измерения")
        if from_cat != to_cat:
            raise ValueError("Единицы из разных категорий")
        cat = from_cat

        if cat == Category.TEMPERATURE:
            result = temp_convert(value, from_unit, to_unit)
        else:
            # Линейная конвертация через базовую единицу
            factor_from = UNITS[cat][from_unit]
            factor_to = UNITS[cat][to_unit]
            in_base = value * factor_from
            result = in_base / factor_to

        # Сохраняем историю
        self.history.append(Conversion(value, from_unit, to_unit, result, cat))
        if len(self.history) > 100:
            self.history = self.history[-100:]
        self.save_config()
        return result, cat

    def display_result(self, value: float, from_unit: str, to_unit: str, result: float, cat: Category):
        print(f"\n{colorize(f'{value:.2f}', Colors.CYAN)} {colorize(from_unit, Colors.YELLOW)} "
              f"{colorize('=', Colors.BOLD)} {colorize(f'{result:.2f}', Colors.GREEN)} {colorize(to_unit, Colors.YELLOW)}")
        print(f"Категория: {colorize(cat.value, Colors.BLUE)}\n")

def interactive(converter: Converter):
    print(colorize("Строительный конвертер единиц (интерактивный режим)", Colors.BOLD))
    print("Введите 'q' для выхода, 'history' для истории")
    while True:
        try:
            value_str = input(colorize("Значение: ", Colors.CYAN))
            if value_str.lower() == 'q':
                break
            if value_str.lower() == 'history':
                for h in converter.history[-10:]:
                    print(f"{h.value:.2f} {h.from_unit} = {h.result:.2f} {h.to_unit} ({h.category.value})")
                continue
            value = float(value_str)
            from_unit = input(colorize("Из единицы: ", Colors.CYAN)).strip().lower()
            if from_unit == 'q': break
            to_unit = input(colorize("В единицу: ", Colors.CYAN)).strip().lower()
            if to_unit == 'q': break
            result, cat = converter.convert(value, from_unit, to_unit)
            converter.display_result(value, from_unit, to_unit, result, cat)
        except ValueError as e:
            print(colorize(f"Ошибка: {e}", Colors.RED))
        except KeyboardInterrupt:
            break

def main():
    parser = argparse.ArgumentParser(description='Construction Unit Converter')
    parser.add_argument('value', nargs='?', type=float, help='Значение для конвертации')
    parser.add_argument('from_unit', nargs='?', help='Исходная единица')
    parser.add_argument('to_unit', nargs='?', help='Целевая единица')
    parser.add_argument('-o', '--output', help='Сохранить результат в файл')
    parser.add_argument('--history', action='store_true', help='Показать историю')
    parser.add_argument('--clear-history', action='store_true', help='Очистить историю')
    args = parser.parse_args()

    converter = Converter()

    if args.clear_history:
        converter.history = []
        converter.save_config()
        print("История очищена.")
        return

    if args.history:
        for h in converter.history[-20:]:
            print(f"{h.value:.2f} {h.from_unit} = {h.result:.2f} {h.to_unit} ({h.category.value})")
        return

    if args.value is not None and args.from_unit and args.to_unit:
        try:
            result, cat = converter.convert(args.value, args.from_unit, args.to_unit)
            converter.display_result(args.value, args.from_unit, args.to_unit, result, cat)
            if args.output:
                with open(args.output, 'a') as f:
                    f.write(f"{args.value} {args.from_unit} = {result} {args.to_unit}\n")
                print(colorize(f"Результат сохранён в {args.output}", Colors.GREEN))
        except Exception as e:
            print(colorize(f"Ошибка: {e}", Colors.RED))
            sys.exit(1)
    else:
        interactive(converter)

if __name__ == '__main__':
    main()

File: converter.ts
// converter.ts
// Версия на TypeScript с классами, интерфейсами, декораторами

import { Command } from 'commander';
import chalk from 'chalk';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import * as readline from 'readline';

// Декоратор для логирования (экспериментальный)
function logMethod(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const original = descriptor.value;
    descriptor.value = function (...args: any[]) {
        console.log(`[LOG] ${propertyKey} вызван с`, args);
        return original.apply(this, args);
    };
    return descriptor;
}

// Типы
type UnitValue = number | string; // для температуры строки
type CategoryMap = Record<string, UnitValue>;

interface UnitCategory {
    [key: string]: CategoryMap;
}

const CATEGORIES: Record<string, CategoryMap> = {
    length: { m: 1, cm: 0.01, mm: 0.001, km: 1000, in: 0.0254, ft: 0.3048, yd: 0.9144, mi: 1609.344 },
    area: { m2: 1, cm2: 0.0001, ft2: 0.09290304, yd2: 0.83612736, acre: 4046.8564224, ha: 10000 },
    volume: { m3: 1, cm3: 1e-6, l: 0.001, ft3: 0.028316846592, yd3: 0.764554857984, gal: 0.003785411784 },
    mass: { kg: 1, g: 0.001, mg: 1e-6, lb: 0.45359237, oz: 0.028349523125, t: 1000 },
    pressure: { pa: 1, kpa: 1000, mpa: 1e6, bar: 100000, psi: 6894.757293168, atm: 101325 },
    temperature: { c: 'celsius', f: 'fahrenheit', k: 'kelvin' },
    angle: { deg: 1, rad: 57.29577951308232, grad: 0.9 },
    speed: { ms: 1, kmh: 0.27777777777778, mph: 0.44704, kn: 0.51444444444444 }
};

interface ConversionEntry {
    value: number;
    fromUnit: string;
    toUnit: string;
    result: number;
    category: string;
}

const CONFIG_FILE = path.join(os.homedir(), '.converter_config.json');

class Converter {
    private history: ConversionEntry[] = [];

    constructor() {
        this.loadConfig();
    }

    private loadConfig(): void {
        try {
            if (fs.existsSync(CONFIG_FILE)) {
                const data = JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
                this.history = data.history || [];
            }
        } catch (e) {}
    }

    @logMethod
    saveConfig(): void {
        fs.writeFileSync(CONFIG_FILE, JSON.stringify({ history: this.history }, null, 2));
    }

    getCategory(unit: string): string | null {
        for (const [cat, units] of Object.entries(CATEGORIES)) {
            if (units[unit] !== undefined) return cat;
            if (cat === 'temperature' && ['c','f','k'].includes(unit)) return cat;
        }
        return null;
    }

    @logMethod
    convert(value: number, fromUnit: string, toUnit: string): { result: number; category: string } {
        const fromCat = this.getCategory(fromUnit);
        const toCat = this.getCategory(toUnit);
        if (!fromCat || !toCat) throw new Error('Неизвестная единица измерения');
        if (fromCat !== toCat) throw new Error('Единицы из разных категорий');
        let result: number;
        if (fromCat === 'temperature') {
            // Функция температуры
            let kelvin: number;
            if (fromUnit === 'c') kelvin = value + 273.15;
            else if (fromUnit === 'f') kelvin = (value + 459.67) * 5/9;
            else if (fromUnit === 'k') kelvin = value;
            else throw new Error(`Неизвестная единица температуры: ${fromUnit}`);
            if (toUnit === 'c') result = kelvin - 273.15;
            else if (toUnit === 'f') result = kelvin * 9/5 - 459.67;
            else if (toUnit === 'k') result = kelvin;
            else throw new Error(`Неизвестная единица температуры: ${toUnit}`);
        } else {
            const factorFrom = CATEGORIES[fromCat][fromUnit] as number;
            const factorTo = CATEGORIES[fromCat][toUnit] as number;
            result = (value * factorFrom) / factorTo;
        }
        this.history.push({ value, fromUnit, toUnit, result, category: fromCat });
        if (this.history.length > 100) this.history = this.history.slice(-100);
        this.saveConfig();
        return { result, category: fromCat };
    }

    display(value: number, fromUnit: string, toUnit: string, result: number, category: string): void {
        console.log(`\n${chalk.cyan(value.toFixed(2))} ${chalk.yellow(fromUnit)} ${chalk.bold('=')} ${chalk.green(result.toFixed(2))} ${chalk.yellow(toUnit)}`);
        console.log(`Категория: ${chalk.blue(category)}`);
        console.log();
    }

    getHistory(): ConversionEntry[] {
        return this.history;
    }

    clearHistory(): void {
        this.history = [];
        this.saveConfig();
    }
}

const program = new Command();
program
    .name('converter')
    .description('Construction Unit Converter (TypeScript)')
    .argument('[value]', 'Значение для конвертации')
    .argument('[from]', 'Исходная единица')
    .argument('[to]', 'Целевая единица')
    .option('-o, --output <file>', 'Сохранить результат в файл')
    .option('--history', 'Показать историю')
    .option('--clear-history', 'Очистить историю')
    .action(async (value, from, to, options) => {
        const converter = new Converter();
        if (options.clearHistory) {
            converter.clearHistory();
            console.log('История очищена.');
            return;
        }
        if (options.history) {
            const hist = converter.getHistory().slice(-20);
            for (const h of hist) {
                console.log(`${h.value.toFixed(2)} ${h.fromUnit} = ${h.result.toFixed(2)} ${h.toUnit} (${h.category})`);
            }
            return;
        }
        if (value !== undefined && from && to) {
            try {
                const val = parseFloat(value);
                const { result, category } = converter.convert(val, from, to);
                converter.display(val, from, to, result, category);
                if (options.output) {
                    fs.appendFileSync(options.output, `${val} ${from} = ${result} ${to}\n`);
                    console.log(chalk.green(`Результат сохранён в ${options.output}`));
                }
            } catch (e: any) {
                console.error(chalk.red(`Ошибка: ${e.message}`));
                process.exit(1);
            }
        } else {
            // Интерактивный режим
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout
            });
            console.log(chalk.bold('Строительный конвертер единиц (интерактивный режим)'));
            console.log('Введите "q" для выхода, "history" для истории');
            const question = (q: string) => new Promise<string>(resolve => rl.question(chalk.cyan(q), resolve));
            while (true) {
                try {
                    let val = await question('Значение: ');
                    if (val.toLowerCase() === 'q') break;
                    if (val.toLowerCase() === 'history') {
                        for (const h of converter.getHistory().slice(-10)) {
                            console.log(`${h.value.toFixed(2)} ${h.fromUnit} = ${h.result.toFixed(2)} ${h.toUnit} (${h.category})`);
                        }
                        continue;
                    }
                    const valueNum = parseFloat(val);
                    if (isNaN(valueNum)) throw new Error('Введите число');
                    const fromUnit = (await question('Из единицы: ')).trim().toLowerCase();
                    if (fromUnit === 'q') break;
                    const toUnit = (await question('В единицу: ')).trim().toLowerCase();
                    if (toUnit === 'q') break;
                    const { result, category } = converter.convert(valueNum, fromUnit, toUnit);
                    converter.display(valueNum, fromUnit, toUnit, result, category);
                } catch (e: any) {
                    console.log(chalk.red(`Ошибка: ${e.message}`));
                }
            }
            rl.close();
        }
    });

program.parse(process.argv);
